migrations/20200717111739-mapping_referance_project_to_private_program.js

```diff
@@ -0,0 +1,111 @@
+var fs = require("fs");
+let _ = require("lodash");
+
+module.exports = {
+  async up(db) {
+
+    global.migrationMsg = "Mapping referance project to private program"
+
+    let usersProjects = await db.collection('userProjects').aggregate([
+      {
+        $match: { "createdType": { $eq: "by reference" } }
+      },
+      {
+        $group: {
+          _id: { userId: { userId: "$userId" } },
+          project: { $push: { _id: "$_id", solutionId: "$solutionId", programId: "$programId" } }
+        }
+      }
+    ]).toArray();
+
+    
+    let usersProjectsArray = _.chunk(usersProjects, 2);
+    async function checkPrivateProgramExist(userId) {
+
+      let programDoc = await db.collection('programs').findOne({ owner: userId }, { _id: 1 });
+      if (programDoc) {
+        return programDoc._id;
+      } else {
+
+        let program = {
+          "externalId": userId + "1",
+          "name": "Improvement Private Program",
+          "description": "Improvement Private Program",
+          "owner": userId,
+          "deleted": false,
+          "isAPrivateProgram": true,
+          "status": "active",
+          "resourceType": [
+            "program"
+          ],
+          "language": [
+            "English"
+          ],
+
+          "keywords": [],
+          "concepts": [],
+          "createdFor": [
+            userId
+          ],
+          "imageCompression": {
+          },
+          components: [],
+        }
+        let programDoc = await db.collection('programs').insertOne(program);
+        return programDoc.insertedId;
+      }
+    }
+
+    try {
+
+      for (let user = 0; user < usersProjectsArray.length; user++) {
+
+        let projects = usersProjectsArray[user].map(userProject => {
+          return userProject;
+        });
+
+        for (let i = 0; i < projects.length; i++) {
+          let project = projects[i];
+
+
+          let userId = project['_id']['userId']['userId'];
+          let userProgramCreation = await checkPrivateProgramExist(userId);
+         
+          let solutionIds = [];
+          let programId;
+         
+          projectIds  = [];
+          project.project.map(document => {
+
+
+            if (!solutionIds.includes(document.solutionId)) {
+              solutionIds.push(document.solutionId);
+            }
+
+            if (!programId) {
+              programId = document.programId;
+            }
+
+            if(!projectIds.includes(document._id)){
+              projectIds.push(document._id);
+            }
+          });
+
+          if (solutionIds && solutionIds.length > 0) {
+            let programUpdate = await db.collection('programs').findOneAndUpdate({ _id: userProgramCreation }, { $push: { components: solutionIds } });
+          }
+          let userProjects = await db.collection('userProjects').updateMany({ createdType: "by reference", userId: userId }, { $set: { programId: userProgramCreation } });
```

Comment:
@rakeshSgr  Delete from userProjects and update in projects.

Reply:
this is related old data migration, In old db we have userProjects only.

Reply:
@rakeshSgr  In dev db we have unnati-redesign database name . Check it and update accordingly.